### ISSUE TITLE
Support compilation testing for the GOcean API

### ISSUE THREAD

**Issue by arporter** [stfc/PSyclone#281]
Although we support compiling generated code (in the test suite) for the dynamo0.3 API, we don't yet have that functionality for GOcean1.0. We will add that here.

**Comment by hiker** [comment]
Quick update:
- I've found one bug in the current handling of f90 flags (a flag like "-I /somewhere" is passed as ONE parameter to the compiler, which then tries to compile "-I /somewhere"). I've already fixed that locally.
- To proper support both APIs some refactoring is necessary. This should also allow us to avoid recompiling the whole infrastructure for each test case. I've already asked via email, but just in case: is anyone else touching test compilations atm?
-I need to supply additional compiler flag(s?) if I want to avoid compiling infrastructure libraries all the time: at least I need to add the include path so that .mod files are found (I _think_ all Fortran compilers will search for .mod files in the include path). Can I just assume that  "-I PATH" will work? Or is there a need to make this more flexible (which would mean to define compiler specific options? E.g. the cray compiler has a "-p" option where module-only search paths can be specified, but it will also search any path specified with -I, so while there could be some use to using -p instead of -I, both will work)

**Comment by arporter** [comment]
I *think* that if you use a fixture then you can get the test system to build the infrastructure just once for all the tests (the 'scope' attribute of a fixture). You then have control of where to put it etc.

**Comment by hiker** [comment]
Atm I am favouring something along the line:
```
    db = Dynamo0p3Build(f90, f90flags, tmpdir)
    assert db.code_compiles(psy)
```
This API-specific compilation flag can then keep track if the infrastructure files have been compiled (in a class variable), and could only be created once per file (in a setup function that's called automatically), or even all the time - its creation is cheap.

**Comment by arporter** [comment]
That looks nice. I've just thought that we might need to be careful about race conditions when running the tests in parallel.

**Comment by hiker** [comment]
Good point, I need to think about this. Even though the compilation has to happen in a fixed order (I don't want to support makefile-like logic in PSyclone), if a file is compiled twice at the same time we could end up with an invalid .mod or .o file. We could add a 'lock' file to the infrastructure dir, but that's kind of messy. Or maybe each process could compile the infrastructure library once in a process-specific directory.

Or worst case - the infrastructure library must be (manually) compiled first :(

**Comment by arporter** [comment]
I think a process-specific directory sounds like the simplest solution.

**Comment by hiker** [comment]
Leaving the parallel compilation to later, one questions:

Is the following code structure ok for the tests:
```
    if TEST_COMPILE:
        # If compilation testing has been enabled (--compile flag to py.test)
        _compile = GOcean1p0Build(f90, f90flags, tmpdir)
        assert _compile.code_compiles(psy)
```
There are a few alternatives we might want to investigate:
- We can do the TEST_COMPILE inside of the compile object, so the 'if TEST_COMPILE' part could be removed from the test files.
- We could (probably) set f90, f90flags, and tmpdir directly into the objects as class variables, so not requiring parameters for compilation at all in the tests (I need to check how exactly fixtures work tbh).
- We could also use GOcean1p0Build as a kind of namespace (i.e. class with static functions), so the code above becomes:
```
   GOcean1p0Build.code_compiles(f90, f90flags, tmpdir, psy)
```
The lines tend to be a bit long this way, but this would be closer to the original code. It might make it easier to change compiler flags if required (atm you create a compile object with a certain flag, so you can use this object for several compiles, without specifying the compiler flags - and I can't decide if this is nice (less parameter in code_compiles), or confusing (no f90flags etc in code_compiles, so you might have to search upwards in the code to find where the flags are specified).

Any preferences? I just want to avoid changing all existing tests to the new structure more than once ;)

**Comment by arporter** [comment]
I would favour moving the `TEST_COMPILE` test down into the infrastructure (I've already done that for the existing dynamo infrastructure but am only modifying the tests themselves when I happen to touch them).
If the `f90` etc. fixtures can be picked-up directly in the compiling class then that would be nice as the list of fixtures tends to get a bit long. (Another alternative is to have a fixture that is itself a dictionary or named tuple that holds all of the different bits of info. required to run a compilation test.)
Personally, I prefer your suggestion of having GOcean1p0Build as a kind of namespace unless there's a performance implication (benefit) with creating the compilation object just once?

**Comment by hiker** [comment]
No real performance impact when using a kind of namespace.

Then I'll do some reading about fixtures and see how I can use them to avoid the parameters completely. That would indeed simplify the tests.

**Comment by hiker** [comment]
OK, after a closer look it found it more elegant to use classes instead of 'namespaces'/static functions (I admit I wasn't sure how without 'self' we could overwrite a function like get_infrastructure_flags)( depending on which class we use). So now it looks like:
```
assert GOcean1p0Build(tmpdir).code_compiles(psy)
assert Dynamo0p3Build(tmpdir).code_compiles(psy)
```
Defining tmpdir in the constructor still allows us to create one compile object and use it more than once if useful, and the syntax is not that different from the original new proposal (Dynamo0p3Build.code_compiles(psy, tmpdir)). There is no need to test for TEST_COMPILE, and f90/f90flags do not need to be passed in to the test, the compile class takes it from the test.py config class.

I still pass in tmpdir to the tests - I couldn't find a good way of avoiding this (since by using the existing fixture we get the benefit of test.py setting-up and cleaning the tmpdir for us, which I think we would otherwise have to do ourselves).

I'll try to quickly convert all of the tests to the new syntax (atm it mostly works since I have provided new functions code_compiles that creates the correct class).

**Comment by hiker** [comment]
@arporter: I'm adding support for OpenCL compilation now. My suggested approach: add a new command line option --compileopencl. If this is specified, all (and only) opencl code will be compiled. It can be combined with --compile (to compile all other code). And it will have the similar symbols TEST_COMPILE_OPENCL and @COMPILE_OPENCL.

Is that ok?

**Comment by arporter** [comment]
That sounds sensible. I did previously have a bit of a think about how to support OpenACC compilation and concluded a new set of flags would be needed. It's not quite the same as OpenCL though because we still use the standard `gen()` method instead of `gen_ocl()`.  Perhaps we'd just run all compilation tests with OpenACC flags specified? Depending on how big this Issue is already, you could perhaps punt OpenCL/OpenACC support to a separate Issue?

**Comment by hiker** [comment]
Yes, re openacc I thought there is no need to handle this especially(?) But OpenCL has at least a 'use fortcl', and fails because of that. The solution of adding a --compileopencl flag is still not idea, e.g.:
```
py.test --compile ./gocean1p0_opencl_test.py
```
This will not actually compile any Fortran files (since they are all opencl tests, and require the --compileopencl flag). Should I print a warning in this case? On the other hand (considering that not many people will have opencl-enabled compiler) this would mean that in most runs useless warnings are printed. Or should we run one test to detect if an opencl enabled compiler is available? That's a bit messy, too.

Oh, I am also trying to get rid of global variables in tests, e.g. the OpenCL variables are part of the Compile base class:
```
from psyclone_test_utils import Compile
@Compile.COMPILE_OPENCL
```
Is that ok? Should I do the same for the normal COMPILE definition (I would opt for consistency over 'no global variables' - either I make COMPILE_OPENCL global as well, or COMPILE a class variable).

**Comment by arporter** [comment]
I think that the best way to ensure that we spot that some tests aren't run is to ensure that one or more of them are *only* run if --compileopencl is set. That way pytest marks them as skipped if the flag is not set which is probably sufficient. We just need to ensure we document this in both the developers' guide and the reviewing guidelines.

I'm happy to get rid of global variables so yes, please do the same for COMPILE as well.

**Comment by arporter** [comment]
Oh, and OpenACC has as runtime API too and so we might sometimes do `use openacc` although I'm not sure that we actually do that in any generated code.

**Comment by hiker** [comment]
> I think that the best way to ensure that we spot that some tests aren't run is to ensure that one or more of them are _only_ run if --compileopencl is set. That way pytest marks them as skipped if the flag is not set which is probably sufficient. We just need to ensure we document this in both the developers' guide and the reviewing guidelines.

Great idea!

> I'm happy to get rid of global variables so yes, please do the same for COMPILE as well.

OK, will do.

> Oh, and OpenACC has as runtime API too and so we might sometimes do use openacc although I'm not sure that we actually do that in any generated code.

I believe at this stage there is no call to a runtime library, since my tests all seem to work so far. But maybe I'll leave this for another ticket, since we might need to look at different compiler flags for different programming models (and we don't want to run all tests four times: normal, openacc, opencl, openmp?). I suggest to leave better support for different compiler options for a separate ticket.

I have parallel compilation for dynamo working. For gocean I am trying a different approach: I just call ''make'' in dl_esm_inf (because I want to avoid hard-coding the compilation order which could be affected by a small change in dl_esm_inf). In parallel pytest runs that means there will be several makes running in the same directory at the same time, which might cause problems. I might see if we
- can risk that, or 
- if I can start the make before the parallel jobs start (problem is that the compile class checks certain command line options like --compile, which are not yet defined if I try to compile in pytest_addoption - but maybe there is a useful later hook that is called before creation of the parallel processes)  or
- if we just document that dl_esm_inf must be compiled before running compile tests with gocean (we could test for the existence of the lib file and raise an error early), or
- or let only child process 0 do the compilation (and hope that no other process has started a compile job .. yes, not a great 'solution' I admit).

Well, work in progress :)

**Comment by hiker** [comment]
I had a good look at what other hooks pytest calls (https://docs.pytest.org/en/latest/reference.html#hooks), and it appears that those hooks are either:
- too early to be able to import a tes class and get it to work
- called once per process when using xdist.

So my proposal for the infrastructure libraries would then be:
- dynamo: will compile the wrapper files one per process (in separate directories)
- gocean: will not compile at all, I am happy to test if the libdl_esm_inf.a file exists and otherwise print a warning (Fortran test compilation requested, but GOcean1p0 infrastructure library is not compile. Please type 'make' in .../external/d_esm_inf/finite_difference). 

Alternatives for gocean would be:
1) run a 'make' in each process (i.e. in parallel). Since there is no easy way to synchronise the pytest processes, that could result in invalid .o/.a files (depends if the compiler tries to handle the case that several processes might write the same .o file)
2) run a 'make' in process 0, but then it is not guaranteed that this 'make' will finish before another process might need a .mod file.
3) We compile once per process in a process-specific directory. While this would solve all of the above issues, the problem is that dl_esm_inf has dependencies between the modules, so we would have to make sure when we update dl_esm_info, to keep the order in which we compile files in sync with the dependencies, otherwise compiles might suddenly fail.

I admit I am happy with either alternative solution, so @arporter, @rupertford, let me know if you are not happy with assuming that dl_esm_info has been compiled previously.

Otherwise I am getting close to a pull request, I just need to check what documentation needs to ne updated.

**Comment by arporter** [comment]
Thanks @hiker. I'd really like to avoid having to have the user compile the infrastructure separately - once per process seems like a better solution Could we simply copy the dl_esm_inf directory into a temporary directory and do the build there? Or, more complicated, alter dl_esm_inf so that it supports having a separate compilation/build directory?

**Comment by hiker** [comment]
20 minutes later, and I would have issued my pull request ... I am just too slow :)
OK, the best solution would be to allow dl_esm_inf to be compiled in a different directory, so I can call 'make' from within the test framework. This would give us all the flexibility of make (i.e. no hard coding of the order in which the files must be compiled), and convenience of not having to pre-compile.

**Comment by arporter** [comment]
PR was merged last month but I forgot to close this issue.
